tracee/tracee.go

```diff
@@ -164,11 +164,24 @@ func (t *Tracee) initBPF() error {
 	}
 	t.bpfModule = bpf.NewModule(string(bpfText), []string{})
 
-	// ensure essential events are traced
-	eventsToTraceInternal := append(t.config.EventsToTrace, essentialEvents...)
-	eventsToTraceInternal = DedupInt32Slice(eventsToTraceInternal)
+	// compile final list of events to trace including essential events while at the same time record which essentials were requested by the user
+	eventsToTraceFinal := make([]int32, 0, len(t.config.EventsToTrace))
+	tmpset := make(map[int32]bool, len(t.config.EventsToTrace))
+	for e := range essentialEvents {
+		eventsToTraceFinal = append(eventsToTraceFinal, e)
+		tmpset[e] = true
+	}
+	for _, e := range t.config.EventsToTrace {
+		_, essential := tmpset[e]
+		if essential {
+			essentialEvents[e] = true // mark that this essential event was requested by the user
+		} else {
+			eventsToTraceFinal = append(eventsToTraceFinal, e)
+		}
+		tmpset[e] = true
```

Comment:
In the previous loop, tmpset contained only essential events.
This is being used in "if essential" above.
Here you add non essential events to tmpset.
So the assumption that tmpset contains only essential events is invalidated after one iteration.

Reply:
there is not assumption that it contains only essential events, but I agree some of the var names like `essential` is misleading. There is another bug here where in the event that `eventsToTrace` contains duplicates, `essentialEvents` map will contain non-essential events. I was working under the assumption that the input is deduped, but I've addressed this concern in: 
https://github.com/aquasecurity/tracee/pull/47/commits/37b85b67c9d97608e2acfc1356767b10453011b8